### ISSUE TITLE
Incorrect QML search path based on import order

### ISSUE THREAD

**Issue by bjones1** [mherrmann/fbs#114]
On OS X Mojave (Python 3.7) with PyQt5 and PySide2 installed, I discovered something interesting: executing the script below via `fbs run` incorrectly reports the QML search path -- it's using PyQt5. Reversing the order of the first two imports fixes the problem.

``` Python
from fbs_runtime.application_context import ApplicationContext
from PySide2.QtQml import QQmlApplicationEngine

import sys

if __name__ == '__main__':
    appctxt = ApplicationContext()
    engine = QQmlApplicationEngine()
    print(engine.importPathList())
```

Output as written: `['/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/Resources/Python.app/Contents/MacOS', 'qrc:/qt-project.org/imports', '/Users/user154992/venv/lib/python3.7/site-packages/PyQt5/Qt/qml']`. **<-- Note the PyQt5!**

Output when the first two lines are reversed (import PySide2 first): `['/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/Resources/Python.app/Contents/MacOS', 'qrc:/qt-project.org/imports', '/Users/user154992/venv/lib/python3.7/site-packages/PySide2/Qt/qml']`. **<-- PySide2 is correct.**

Version info (editing to show only useful stuff):

```
(venv) 74-80-250-229:fbs_test user154992$ pip list
Package               Version            Location                     
--------------------- ------------------ -----------------------------
fbs                   0.8.2              
PyQt5                 5.12.2             
PyQt5-sip             4.19.17            
PySide2               5.12.3             
shiboken2             5.12.3             
```

**Comment by mherrmann** [comment]
Interesting, I've also experienced problems when both PyQt5 and PySide2 are installed. fbs's runtime [mentions both](https://github.com/mherrmann/fbs/blob/1327c4d5028b44ca949bdec72098e930bf79fae3/fbs_runtime/application_context.py#L8-L15). I suspect that this makes PyInstaller package them both. Maybe whichever one is imported first, like you showed, then decides them QML search path.

Correct me if I'm wrong, but neither PyInstaller nor fbs can (really) know which of the two the user is actually using.

One workaround would be to require the user to only have one of the two bindings installed. Another would be to introduce a special flag in fbs that decides which binding is used.

**Comment by bjones1** [comment]
The link is helpful -- thanks! You're correct that PyInstaller will package both, since it does static (dumb) analysis of all imports. As you say, one solution I can see it instruct users to only install one in the venv they're building from. Is this already in the docs? I've browsed, but not really read them. You've done a great job of documenting fbs!

You're correct that PyInstaller/fbs can't know *a priori* which to use. Another solution might be to have users `from fbs_runtime.PyQt5.application_context import ApplicationContext` or something like that. That's my favorite approach. I'd suggest avoiding a special flag in a JSON file, since I think that's more obscure and harder to implement.

**Comment by bjones1** [comment]
I did some testing on my (cloud-based) Mac. Just uninstalling PyQt5 didn't work -- PyInstaller's PyQt5 hooks break. I'll need to do some thinking on how to nicely fix that.

As a workaround, I removed all PyQt5 imports from fbs (in application_context and, I think, signal?). That fixed the error -- my test app above now works correctly.

**Comment by mherrmann** [comment]
> As you say, one solution I can see it instruct users to only install one in the venv they're building from. Is this already in the docs?

Not yet, no. It's more a bug that there are problems if the user has both PyQt5 and PySide2.

> You've done a great job of documenting fbs!

Thank you! I'm happy to hear you like the docs.

> Another solution might be to have users `from fbs_runtime.PyQt5.application_context import ApplicationContext` or something like that. 

That's actually very elegant.

> I'd suggest avoiding a special flag in a JSON file, since I think that's more obscure and harder to implement.

Good point.

> As a workaround, I removed all PyQt5 imports from fbs (in application_context and, I think, signal?). That fixed the error

Then I suspect the import solution you suggested should also work to fix the issue. Wouldn't this fix all of our current problems?

**Comment by bjones1** [comment]
Yes, I believe it would. I could suggest an approach on implementing this if you'd like.

**Comment by mherrmann** [comment]
Thank you! But let me think a little about what it means fbs first, before I make you do unnecessary work.

**Comment by mherrmann** [comment]
Is there a way to make PyInstaller ignore imports mentioned in a particular file / module? Then we could have an even nicer solution: I could add those fbs modules that mention PyQt5/PySide2 to that list. And which binding is used would be wholly determined by the user's imports.

**Comment by mherrmann** [comment]
(Crucially, this approach would even let us avoid the `from fbs.PyQt5...` workaround you mentioned.)

**Comment by bjones1** [comment]
While there is an [--exclude-module](https://pyinstaller.readthedocs.io/en/stable/usage.html#what-to-bundle-where-to-search) option, I don't think it will work. This flag applies globally, to the entire program, not just to fbs. So, it would exclude PyQt5 and PySide2 from the final executable. Also, the exclude is a per-module option, so to make it work you'd need to exclude **every** PyQt5/PySide2 import, which is hard to do.

**Comment by mherrmann** [comment]
I see :-) Well, I'm sure there are tricks one can use to avoid PyInstaller picking up the import. Eg instead of

    from PyQt5 import QtWidgets

one may be able to write

    from importlib import import_module
    QtWidgets = import_module('PyQt5.QtWidgets')

**Comment by mherrmann** [comment]
I'll look into it.

**Comment by bjones1** [comment]
Correct -- that would work. However, in the non-frozen case, this still means both PySide2 and PyQt5 get imported, which is IMHO a bad thing.

**Comment by mherrmann** [comment]
Again, you're right. If the user wants to use PySide2 then fbs's [`application_context.py`](https://github.com/mherrmann/fbs/blob/1327c4d5028b44ca949bdec72098e930bf79fae3/fbs_runtime/application_context.py#L8-L15) still imports - and uses! - PyQt5's classes. Hm... :thinking:

**Comment by bjones1** [comment]
I was thinking something like the following:

- `fbs_runtime/`
  - `application_context/`
  - `__init___.py` (blank)
  - `PyQt5.py`
  - `PySide2.py`
  - `_core.py`

In `_core.py`:

``` Python
# A place to put PyQt5/PySide2 classes
class _PyQt:
    pass

# The rest of the contents are from the old application_context.py. 
# Instead of using QApplication, use _PyQt.QApplication.
# Instead of using QIcon, use _PyQt.QIcon.
```

In `PyQt5.py`:

``` Python
from PyQt5.QtGui import QIcon
from PyQt5.QtWidgets import QApplication

import _application_context
from _core import *

# Tell the core which flavor of QIcon, QApplication to use.
_core._PyQt.QIcon = QIcon
_core._PyQt.QApplication = QApplication
```

Similarly for `PySide2.py`.

Just an idea...untested...

**Comment by mherrmann** [comment]
I really appreciate your thoughts. My feeling would be more something along these lines:

 * `fbs_runtime/`
   * `application_context/`
      * `__init__.py`
      * `pyqt5.py`
      * `pyside2.py`

In `application_context/__init__.py`:

```
class _ApplicationContext:
    @cached_property
    def qt_binding(self):
        raise NotImplementedError()
    @cached_property
    def app(self):
        result = self.qt_binding.QApplication([])
        ...
    ...

_QtBinding = namedtuple('_QtBinding', ('QApplication', 'QIcon', 'QAbstractSocket'))
```

In `application_context/pyqt5.py`:

```
from . import _ApplicationContext, _QtBinding
from PyQt5.QtGui import QIcon
from PyQt5.QtWidgets import QApplication
from PyQt5.QtNetwork import QAbstractSocket

class PyQt5ApplicationContext(_ApplicationContext):
    @cached_property
    def qt_binding(self):
        return _QtBinding(QApplication, QIcon, QAbstractSocket)
```

Users would then write:

    from fbs_runtime.application_context.pyqt5 import PyQt5ApplicationContext

What do you think of this API / these names?

**Comment by bjones1** [comment]
I like it! Very nice. I'd suggest using `PyQt5` instead of `pyqt5`, though.

**Comment by mherrmann** [comment]
Thanks. Why do you prefer `PyQt5` over `pyqt5`? I think I'm subconsciously trying to emphasize that my modules aren't `PyQt5`; They're only related.

**Comment by bjones1** [comment]
Since I use PyQt5 everyone else :). It just feels odd to omit the capitals.

Personally, I'd also use `from fbs_runtime.application_context.PyQt5 import ApplicationContext` since it's a bit DRYer.

Just my taste, though.

**Comment by mherrmann** [comment]
Okay. I'll follow your suggestion :slightly_smiling_face: 

I'm in the process of changing fbs's code. While testing, PyInstaller still seems to package PyQt5 even if none of fbs's code mentions it. But it is installed in my venv. Could it be that I have to install a snapshot version of PyInstaller?

**Comment by bjones1** [comment]
Oops -- sorry, I forgot to mention a fixed version I'm testing at https://github.com/bjones1/pyinstaller/tree/mac-qt-fix. However, it should only affect the way PyInstaller creates Mac .app bundles, not what's packaged.

**Comment by mherrmann** [comment]
That link is dead, unfortunately. When fbs invokes PyInstaller, a file `graph-MyApp.dot` is produced. This seems to indicate the cross-module dependencies. The interesting one is that this graph only mentions PyInstaller's own [`pyi_rth_qt5plugins.py`](https://github.com/pyinstaller/pyinstaller/blob/v3.4/PyInstaller/loader/rthooks/pyi_rth_qt5plugins.py). This apparently handles Qt5 plugins for both PySide2 and for PyQt5. My sample app uses PySide2. My suspicion is that this triggers `pyi_rth_qt5plugins.py`, which in turn makes PyInstaller include both `PySide2` _and_ `PyQt5` when freezing my app. This hypothesis is compatible with the fact that the superfluous PyQt5 modules I get are `QtCore` and `sip`, which are precisely what's mentioned in `pyi_rth_qt5plugins.py`.

Do you know if my hypothesis could be correct?

**Comment by bjones1** [comment]
I just merged my code into the official PyInstaller repo -- sorry, I should have posted here. It's now available via `pip install https://github.com/pyinstaller/pyinstaller/archive/develop.zip` as usual.

You're using an older (buggier) version of PyInstaller -- the current develop has https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/loader/rthooks/pyi_rth_pyqt5.py and https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/loader/rthooks/pyi_rth_pyside2.py instead. Would you mind testing with develop instead?

**Comment by mherrmann** [comment]
Thanks! That works now :-) I released [fbs v0.8.3](https://github.com/mherrmann/fbs/releases/tag/v0.8.3) that incorporates the fix we discussed. Thank you again for your valuable help. It really lead to a better outcome.

**Comment by bjones1** [comment]
Fantastic! Thanks for your help -- I was also able to fix a few bugs in PyInstaller I hadn't noticed before.
